### PR TITLE
impl EndpointError for Error

### PR DESCRIPTION
**Pull request by DevinR528**
Serialization side of [ruma-api issue 37](https://github.com/ruma/ruma-api/issues/37).

### CODE REVIEW

**Comment by DevinR528** [comment]
Am I missing something 
<details close>
<summary>This compiler error</summary>

```
error[E0277]: the trait bound `ruma_api::error::DeserializationError: std::convert::From<strum::ParseError>` is not satisfied
   | 
  ::: /home/devinr/.cargo/git/checkouts/ruma-api-eb55b09b2e2a3fb3/6a0ab98/src/error.rs:88:21
   |
88 |            inner: impl Into<DeserializationError>,
   |                        -------------------------- required by this bound in `ruma_api::error::RequestDeserializationError::new`
  --> src/r0/receipt/create_receipt.rs:9:1
   |
9  |  / ruma_api! {
10 |  |     metadata {
11 |  |         description: "Send a receipt event to a room.",
12 |  |         method: POST,
...   |
31 |  |     response {}
32 |  | }
   |  | ^
   |  | |
   |  |_the trait `std::convert::From<strum::ParseError>` is not implemented for `ruma_api::error::DeserializationError`
   |    in this macro invocation
   | 
  ::: /home/devinr/.cargo/git/checkouts/ruma-api-eb55b09b2e2a3fb3/6a0ab98/ruma-api-macros/src/lib.rs:29:1
   |
29 | /  pub fn ruma_api(input: TokenStream) -> TokenStream {
30 | |      let raw_api = parse_macro_input!(input as RawApi);
31 | |      match Api::try_from(raw_api) {
32 | |          Ok(api) => api.into_token_stream().into(),
33 | |          Err(err) => err.to_compile_error().into(),
34 | |      }
35 | |  }
   | |__- in this expansion of `ruma_api!`
   |
   = help: the following implementations were found:
             <ruma_api::error::DeserializationError as std::convert::From<error::_IMPL_DESERIALIZE_FOR_ErrorKind::_serde::de::value::Error>>
             <ruma_api::error::DeserializationError as std::convert::From<ruma_identifiers::error::Error>>
             <ruma_api::error::DeserializationError as std::convert::From<serde_json::error::Error>>
             <ruma_api::error::DeserializationError as std::convert::From<std::convert::Infallible>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<ruma_api::error::DeserializationError>` for `strum::ParseError`

error: aborting due to 8 previous errors
```
</details>

is implemented [here](https://github.com/ruma/ruma-api/blob/master/src/error.rs#L256)?

**Comment by jplatte** [comment]
The error comes from the versions of strum now mismatching between ruma-api and ruma-client-api (I updated it [here](https://github.com/ruma/ruma-api/commit/6f5e25cb7d9c05a54df4df615773ba048cd945b4)). You need to update it here to match the change in ruma-api.

**Comment by DevinR528** [comment]
Is this moving in the right direction? Should I use the `Error` for some of `Request`/`Response` delcarations?

**Comment by jplatte** [comment]
Only the `status_code` field is meant to be filled from the status code. The other fields are meant to be filled from the response body (see https://matrix.org/docs/spec/client_server/r0.6.0#api-standards).

File: src/error.rs [hunk 1]
```diff
@@ -114,7 +116,19 @@ pub struct Error {
     /// A human-readable error message, usually a sentence explaining what went wrong.
     pub message: String,
     /// The http status code
-    pub status_code: http::StatusCode,
+    pub status_code: StatusCode,
+}
+
+impl EndpointError for Error {
+    fn try_from_response(
+        response: ruma_api::exports::http::Response<Vec<u8>>,
```
Comment (jplatte) [inline src/error.rs]: `ruma_api::exports` is only meant to be used from ruma-api-macros.

```suggestion
        response: http::Response<Vec<u8>>,
```

File: src/error.rs [hunk 2]
```diff
@@ -114,7 +116,19 @@ pub struct Error {
...
+    ) -> Result<Self, ResponseDeserializationError> {
+        if let Ok(error_body) = from_slice::<ErrorBody>(response.body()) {
+            Ok(error_body.into_error(response.status()))
+        } else {
```
Comment (jplatte) [inline src/error.rs]: You're throwing away the error message of `from_slice` here. Also, you're using the nominally private `from_response`. I guess for this to work, at least the other constructor should be public.

File: src/unversioned/get_supported_versions.rs [hunk 3]
```diff
@@ -23,4 +23,7 @@ ruma_api! {
         #[serde(default, skip_serializing_if = "HashMap::is_empty")]
         pub unstable_features: HashMap<String, bool>
     }
+
+    error: crate::error::Error
+
```
Comment (jplatte) [inline src/unversioned/get_supported_versions.rs]: Please remove this extra newline.

File: src/error.rs [hunk 4]
```diff
@@ -114,7 +116,19 @@ pub struct Error {
...
     /// The http status code
-    pub status_code: http::StatusCode,
+    pub status_code: StatusCode,
+}
+
+impl EndpointError for Error {
+    fn try_from_response(
+        response: ruma_api::exports::http::Response<Vec<u8>>,
+    ) -> Result<Self, ResponseDeserializationError> {
+        if let Ok(error_body) = from_slice::<ErrorBody>(response.body()) {
```
Comment (jplatte) [inline src/error.rs]: Style nit: Since `from_slice` is a very generic name and it's not immediately obvious from context, please don't import it into scope and instead refer to it with its full path.

```suggestion
        if let Ok(error_body) = serde_json::from_slice::<ErrorBody>(response.body()) {
```

File: src/error.rs [hunk 5]
```diff
@@ -1,6 +1,8 @@
 //! Errors that can be sent from the homeserver.
-
```
Comment (jplatte) [inline src/error.rs]: Please don't remove the empty line between the module doc comment and the imports.

File: src/error.rs (hunk 4 above)
Comment (DevinR528) [reply src/error.rs]: Yea I looked around a bit and all I could find were uses of this in tests that imported it directly but I agree its too vague.

File: src/error.rs [hunk 6]
```diff
@@ -121,12 +121,11 @@ pub struct Error {
 
 impl EndpointError for Error {
     fn try_from_response(
-        response: ruma_api::exports::http::Response<Vec<u8>>,
+        response: Response<Vec<u8>>,
```
Comment (jplatte) [inline src/error.rs]: Please use `http::Response` as suggested before and like in other parts of the code. There's lots of types called `Response` in ruma-client-api.

**Comment by DevinR528** [comment]
Can I switch the Cargo.toml file back to using `ruma-api = "0.14.?` instead of `ruma-api = { git = "github.com/ruma/ruma-api" }`

**Comment by jplatte** [comment]
No, then ruma-client-api would fail to build. I think I will release a dev version of ruma-api tomorrow which ruma-client-api can then depend on until a regular release is made (or I might do a regular release).

**Comment by jplatte** [comment]
Please resolve the merge conflicts and bump `ruma-api` to `0.15.0-dev.1`.

File: src/error.rs (hunk 6 above)
Comment (jplatte) [reply src/error.rs]: @DevinR528 I seem to remember you changing this as suggested, but that has now been reverted apparently?

File: src/error.rs [hunk 7]
```diff
@@ -114,7 +116,18 @@ pub struct Error {
     /// A human-readable error message, usually a sentence explaining what went wrong.
     pub message: String,
     /// The http status code
-    pub status_code: http::StatusCode,
+    pub status_code: StatusCode,
```
Comment (jplatte) [inline src/error.rs]: Please undo this unrelated change.

**Comment by jplatte** [comment]
Also if you could write a changelog entry that would be nice (you can say ruma-api was updated to 0.15.0, since that's going to have to be released before this crate's next release).

File: CHANGELOG.md [hunk 8]
```diff
@@ -13,7 +13,7 @@ Improvements:
 * Add `filter` type for `r0::context::get_context`
 
 Breaking changes:
-
+* Update ruma-api to 0.15.0
```
Comment (jplatte) [inline CHANGELOG.md]:
```suggestion

* Update ruma-api to 0.15.0
```

Comment (DevinR528) [reply CHANGELOG.md]: hows this
`* response deserialization now takes the above `Error` type into account (introduced in ruma-api 0.15.0)`

Comment (jplatte) [reply CHANGELOG.md]: maybe

```
  * the response deserialization code will try to create an instance of this type from http responses that indicate an error
```

Comment (jplatte) [reply CHANGELOG.md]:
```suggestion
* Update ruma-api to 0.15.0
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -14,6 +15,7 @@ Improvements:
 
 Breaking changes:
 
+* Update ruma-api to 0.15.0
 * Update ruma-api to 0.14.0
```
Comment (jplatte) [inline CHANGELOG.md]:
```suggestion
* Update ruma-api to 0.15.0
```
